Inject a laser from an E(x,y,t) file

lucafedeli88 opened the pull request:
### Descrption
This PR adds a new laser profile: the  field is injected reading the spatio-temporal profile from disk.

The user has to write something like the following in the inputfile:
```
#################################
lasers.nlasers      = 1
lasers.names        = laser1
laser1.position     = 0. 0. 0.e-6 # This point is on the laser plane
laser1.direction    = 1. 0. 0.     # The plane normal direction
laser1.polarization = 0. 0. 1.     # The main polarization vector
laser1.e_max        = 4.e12        # Maximum amplitude of the laser field (in V/m)
laser1.wavelength = 0.8e-6         # The wavelength of the laser (in meters)
laser1.profile      = from_txye_file
laser1.txye_file_name = "las2d.txye"
laser1.time_chunk_size = 100 #(optiona)l
```

`las2d.txye` is a binary file with the following format:
- a flag to indicate if the grid is uniform or not (1 byte, 0 means non-uniform, !=0 means uniform)
- nt, number_of_timesteps (uint32_t, , must be >=2)
- nx, number_of_x_coords (uint32_t, , must be >=2)
- ny, number_of_y_coords (uint32_t, must be 1 for 2D simulations and >=2 for 3D simulations)
- timesteps (double[2] if grid is uniform, double[np] otherwise)
- x_coords (double[2] if grid is uniform, double[nx] otherwise)
- y_coords (double[1] if 2D, double[2] if 3D & uniform grid, double[ny] if 3D & non uniform grid)
- field_data (double[double[nt * nx * ny], with nt being the slowest coordinate],
with number_of_timesteps being the slowest coordinate)

The spatiotemporal grid must be rectangular, but it can be non-uniform. If the grid is non-uniform the code uses a binary search to find the nearest grid points. The field amplitude is calculated with a bilinear (in 2D) or trilinear (in 3D) interpolation. 

Since the file can be very large, the user can set the `time_chunk_size` parameter: only `time_chunk_size` timesteps are read from disk when this is necessary.  

### Changes

- `Source/Laser/LaserProfiles.H` I've added the new laser profile to the list. I've also added a new method to the `ILaserProfile` interface: `update(simulation_time)`. This method is called before fill-amplitude and should be used to update internal parameters of the laser profile (e.g. load new data from disk). Finally, I've added `const` qualifier to some methods. 
- `Source/Laser/LaserParticleContainer.cpp` `update(simulation_time)` is now called before filling the amplitude
- ` Source/Laser/LaserProfilesImpl/LaserProfileGaussian.cpp `,  ` Source/Laser/LaserProfilesImpl/LaserProfileHarris.cpp `  ` Source/Laser/LaserProfilesImpl/LaserProfileFieldFunction.cpp ` minor changes
- `Source/Utils/WarpXUtil.H/cpp` I've added some GPU-friendly routines to perform linear, bilinear and trilinear interpolation
- ` Source/Laser/LaserProfilesImpl/LaserProfileFromTXYEFile.cpp ` This file contains the implementation of the new laser profile
-` Source/Laser/LaserProfilesImpl/Make.package` added line to include ` Source/Laser/LaserProfilesImpl/LaserProfileFromTXYEFile.cpp ` in the compilation
-  `Docs/source/running_cpp/parameters.rst` I've documented this new feature
- `Examples/Tests/laser_injection_from_file/analysis.py` `Examples/Tests/laser_injection_from_file/inputs.2d_test_txye ` `Regression/WarpX-tests.ini ` I've written a draft of a unit test for this feature **help needed!**

### Residual Issues

- [ ] I am not sure about how to write a proper test for this feature
- [ ] I've experienced a segfault using Intel compiler and KNL processors

MaxThevenet commented:
Thanks a lot for this PR @lucafedeli88! I will look at it soon, and I can help you build an automated test, but here are a few preliminary questions:
- Have you tested this feature with mesh refinement (for instance with a simulation where the antenna intersects a refined patch)? Have you tried it with and without sub-cycling? If not, `Examples/Tests/laser_on_fine/inputs.2d` could be a good starting point.
- Have you tested this feature with boosted frame simulations?
- As a future plan (not in this PR), would it be reasonable to try and use the OpenPMD standard for the input format? Note that I am not sure it is possible, as the OpenPMD standard may not support non-uniform grids. @RemiLehe @ax3l any comment?

I can help with these tests if needed. Sorry if these questions were asked and answered earlier, I may have missed some recent discussions.

lucafedeli88 commented:
Thanks for your help! 

>     * Have you tested this feature with mesh refinement (for instance with a simulation where the antenna intersects a refined patch)? Have you tried it with and without sub-cycling? If not, `Examples/Tests/laser_on_fine/inputs.2d` could be a good starting point.

I've just done that (mesh refinement + sub cycling) and it seems to be working:
![mesh_ref_w_sub](https://user-images.githubusercontent.com/4558310/70791126-bf4cc200-1d96-11ea-9fd1-80a67b6f40aa.png)
The antenna is in the middle of the box with a normal equal to [1,1]/sqrt(2). The field file contains a Gaussian beam (see `Example/Tests/laser_injection_from_file/analysis.py`)

>     * Have you tested this feature with boosted frame simulations?

Not yet.

>     * As a future plan (not in this PR), would it be reasonable to try and use the OpenPMD standard for the input format? Note that I am not sure it is possible, as the OpenPMD standard may not support non-uniform grids. @RemiLehe @ax3l any comment?

I would really like to use a standard format such as OpenPMD! I am not a big fan of custom binary file formats, since parsing a file is very error-prone... Using a dedicated library is clearly the best thing to do. I am not familiar with the OpenPMD format though... 

> I can help with these tests if needed. Sorry if these questions were asked and answered earlier, I may have missed some recent discussions.

I would really appreciate some help on this point! I have started preparing a test which first runs a script to prepare a binary field file for a Gaussian beam (`Example/Tests/laser_injection_from_file/analysis.py` & `Regression/WarpX-tests.ini`). However, I am not sure about how to test the quality of the injected Gaussian beam. A comparison between the expected field (red points) along the beam axis and the field retrieved from the simulation (blue points) gives apparently a fair agreement:
![gaussians_comp](https://user-images.githubusercontent.com/4558310/70792054-b3fa9600-1d98-11ea-8c85-6e762d845fb5.png)
However, if I do a point to point difference, I get a low (few % discrepancy) only if I shift the expected beam by roughly 1.5*dt:
![gaussians_comp_2](https://user-images.githubusercontent.com/4558310/70792214-08057a80-1d99-11ea-9d9e-d6a3fa899add.png)
I can't find any reasonable explanation for that...

MaxThevenet commented:
Here's a suggestion for the test:
- Compute the FFT and make sure the spectrum is close to what's expected (typically that the central frequency  corresponds to the input).
- Compute the Hilbert transform (which can give you the laser envelope) and make sure than the full 2D envelope corresponds to the input.

That should be enough, and I can work on that.

MaxThevenet commented:
Hi Luca,

I did a little correctness test for your PR, just for the envelope. For the sake of simplicity, the antenna radiates along z and this uses the CKC solver, that required small changes in this new
[input file](https://github.com/ECP-WarpX/WarpX/files/3963068/inputs.2d_test_txye.txt). The test is done in this [Python script](https://github.com/ECP-WarpX/WarpX/files/3963071/script.py.txt).

Would that help you design an automated test? Note that this only tests the envelope, which is (when computed with the Hilbert transform) kind of independent on the wavelength, so we should probably add a test on the laser wavelength (checking the max of the FFT of the fields, for instance).

lucafedeli88 commented:
Thanks really a lot! I am working on the automated test right now...

lucafedeli88 commented on Examples/Tests/laser_injection_from_file/inputs.2d_test_txye:
```diff
@@ -0,0 +1,46 @@
+interpolation.nox = 3
+interpolation.noy = 3
+interpolation.noz = 3
+warpx.cfl = 1.0
```
sure!

lucafedeli88 commented on Examples/Tests/laser_injection_from_file/analysis.py:
```diff
@@ -0,0 +1,192 @@
+#!/usr/bin/python3
```
I agree, I've just moved it in the folder that you have suggested

lucafedeli88 commented on Docs/source/running_cpp/parameters.rst:
```diff
@@ -538,6 +538,36 @@ Laser initialization
+      values. x,y,t are meant to be in S.I. units, while the field value is meant to be multiplied by
+      ``<laser_name>.e_max`` (i.e. in most cases the maximum of abs(E(x,y,t)) should be 1,
+      so that the maximum field intensity can be set straightforwardly with ``<laser_name>.e_max``).
+      The binary file has to respect the following format:
```
Thanks for the advice, I've just added this to the documentation...

lucafedeli88 commented on Examples/Tests/laser_injection_from_file/analysis.py:
```diff
@@ -0,0 +1,192 @@
+
+#ADD COMMENT
+#ADD COMMENT
+#ADD COMMENT
```
Thanks for this suggestion! I've just added it.

lucafedeli88 commented on Examples/Tests/laser_injection_from_file/analysis.py:
```diff
@@ -0,0 +1,192 @@
+
+# Function for the envelope
+def gauss_env(T,XX,ZZ):
+
```
Thanks, I've just added this comment

lucafedeli88 commented on Examples/Tests/laser_injection_from_file/analysis.py:
```diff
@@ -0,0 +1,192 @@
+tcoords = np.linspace(t_l, t_r, t_points)
+xcoords = np.linspace(x_l, x_r, x_points)
+
+def gauss(T,X,Y,opt):
```
thanks, I've just added this comment

lucafedeli88 commented on Examples/Tests/laser_injection_from_file/analysis.py:
```diff
@@ -0,0 +1,192 @@
+    exp_arg = - (X*X)*inv_w_2 - inv_tau2 / c/c * (Z-T*c)*(Z-T*c)
+    return E_max * np.real(np.exp(exp_arg))
+
+def write_file(fname, x, y, t, E):
```
I've just added your suggestion

lucafedeli88 commented on Examples/Tests/laser_injection_from_file/analysis.py:
```diff
@@ -0,0 +1,192 @@
+        file.write(E.tobytes())
+
+
+def write_file_unf(fname, x, y, t, E):
```
Thanks for your suggestion. I've added a complete description, similar to that of `write_file` . I've also fixed a small bug in the function.

lucafedeli88 commented on Source/Laser/LaserProfiles.H:
```diff
@@ -60,6 +62,17 @@ public:
+     * Some laser profiles might need to perform an "update" operation per
+     * time step.
+     *
+     * @param[in] t simulation time (seconds)
```
thanks!

lucafedeli88 commented on Source/Utils/WarpXUtil.H:
```diff
@@ -16,13 +16,88 @@ void NullifyMF(amrex::MultiFab& mf, int lev, amrex::Real zmin,
+        return first;
+}
+
+/* \brief Performs a linear interpolation
```
Thanks

lucafedeli88 commented on Source/Laser/LaserProfilesImpl/LaserProfileFromTXYEFile.cpp:
```diff
@@ -0,0 +1,534 @@
+/** \brief Reads new field data chunk from file if needed
+ *
+ * @param[in] t simulation time (seconds)
+ */
```
Sure

lucafedeli88 commented on Source/Laser/LaserProfilesImpl/LaserProfileFromTXYEFile.cpp:
```diff
@@ -0,0 +1,534 @@
+ * \param Yp: pointer to second component of positions of laser particles
+ * \param t: Current physical time
+ * \param amplitude: pointer to array of field amplitude.
+ */
```
sure

lucafedeli88 commented on Source/Laser/LaserProfilesImpl/LaserProfileFromTXYEFile.cpp:
```diff
@@ -0,0 +1,534 @@
+ * The spatiotemporal grid must be rectangular.
+ *
+ * \param txye_file_name: name of the file to parse
+ */
```
sure

lucafedeli88 commented on Source/Laser/LaserProfilesImpl/LaserProfileFromTXYEFile.cpp:
```diff
@@ -0,0 +1,534 @@
+    int t_sizes[6] = {m_params.nt, m_params.nx, m_params.ny,
+        static_cast<int>(m_params.t_coords.size()),
+        static_cast<int>(m_params.x_coords.size()),
+        static_cast<int>(m_params.y_coords.size())};
```
When a uniform grid is used `t_coords`, `x_coords` (and `y_coords` in 3D) only store 2 values (the extremes of the mesh). nt, nx (and ny in 3D) store the number of points used for the mesh... I admit that the name choice is a bit misleading...

MaxThevenet replied:
Thanks!
I am fine with keeping the names and add some comments, both here and in the header file (see other comment).

lucafedeli88 replied:
I've added some comments to explain this

lucafedeli88 commented on Source/Laser/LaserProfilesImpl/LaserProfileFromTXYEFile.cpp:
```diff
@@ -0,0 +1,534 @@
+ * \param Yp: pointer to second component of positions of laser particles
+ * \param t: Current physical time
+ * \param amplitude: pointer to array of field amplitude.
+ */
```
sure

lucafedeli88 commented on Source/Laser/LaserProfilesImpl/LaserProfileFromTXYEFile.cpp:
```diff
@@ -0,0 +1,534 @@
+ *
+ * \param t_begin: left limit of the timestep range to read
+ * \param t_end: right limit of the timestep range to read (t_end is not read)
+ */
```
sure

lucafedeli88 commented on Source/Laser/LaserProfilesImpl/LaserProfileFromTXYEFile.cpp:
```diff
@@ -0,0 +1,534 @@
+ * \param Yp: pointer to second component of positions of laser particles
+ * \param t: Current physical time
+ * \param amplitude: pointer to array of field amplitude.
+ */
```
sure

MaxThevenet commented on Source/Laser/LaserProfiles.H:
```diff
@@ -179,6 +204,138 @@ private:
+        int first_time_index; //Index of the first timestep in memory
+        int last_time_index; //Index of the last timestep in memory
+        amrex::Gpu::ManagedVector<amrex::Real> E_data; //Field data
+    } m_params;
```
I am fine with keeping the code as is, and add documentation to avoid confusion.
Also, let's make sure the documentation can be read by Doxygen (as an example the two links below show the Doxygen-generated documentation for class ILaserProfile on `dev` branch, let's make sure Doxygen also reads members of `m_params`).
https://warpx.readthedocs.io/en/latest/_static/doxyhtml/class_warp_x_laser_profiles_1_1_i_laser_profile.html
https://warpx.readthedocs.io/en/latest/api/class_warp_x_laser_profiles_1_1_i_laser_profile.html

lucafedeli88 replied:
Wow, the documentation is awesome!

lucafedeli88 replied:
I've just added the comments...

lucafedeli88 commented:
> Thanks for the changes! Concerning your concerns about clarifying `x_coords` etc. I think this could be done by simply adding comments (see suggestions). After these small changes, I think the PR will be ready to be merged :).

Thanks for your review! I've added the comments, but for unknown reasons I am not able to build the documentation anymore (not even on the dev branch). So I don't know if what I've added can be parsed by Doxygen...
